services/sqs/src/main/java/software/amazon/awssdk/services/sqs/internal/batchmanager/ReceiveQueueBuffer.java

```diff
@@ -0,0 +1,192 @@
+            return;
+        }
+
+        queueAttributesManager.getVisibilityTimeout().thenAcceptAsync(visibilityTimeoutNanos -> {
+            int max = Math.max(config.maxInflightReceiveBatches(), 1);
+            int toSpawn = max - inflightReceiveMessageBatches.get();
+            if (toSpawn > 0) {
+                AsyncReceiveMessageBatch asyncReceiveMessageBatch = new AsyncReceiveMessageBatch(
+                    queueUrl, sqsClient, visibilityTimeoutNanos, config);
+                inflightReceiveMessageBatches.incrementAndGet();
+                asyncReceiveMessageBatch.asyncReceiveMessage()
+                                        .whenComplete((response, exception) -> reportBatchFinished(response));
+            }
+        });
+    }
+
+    private int determineDesiredBatches() {
+        int desiredBatches = Math.max(config.maxDoneReceiveBatches(), 1);
+
+        if (config.adaptivePrefetching()) {
+            int totalRequested = futures.stream()
+                                        .mapToInt(ReceiveMessageCompletableFuture::getRequestedSize)
+                                        .sum();
+            int batchesNeededToFulfillFutures = (int) Math.ceil((float) totalRequested / config.maxBatchItems());
+            desiredBatches = Math.min(batchesNeededToFulfillFutures, desiredBatches);
+        }
+
+        return desiredBatches;
+    }
+
+    private void fulfillFuture(ReceiveMessageCompletableFuture future, AsyncReceiveMessageBatch task) {
+        List<Message> messages = new LinkedList<>();
+        Throwable exception = task.getException();
+        int numRetrieved = 0;
+        boolean batchDone = false;
+
+        if (exception != null) {
+            future.setFailure(exception);
+            finishedTasks.poll();
+            return;
+        }
+
+        while (numRetrieved < future.getRequestedSize()) {
+            Message msg = task.removeMessage();
+            if (msg != null) {
+                messages.add(msg);
+                ++numRetrieved;
+            } else {
+                batchDone = true;
+                break;
+            }
+        }
+        batchDone = batchDone || task.isEmpty();
+        if (batchDone) {
+            lock.lock();
+            try {
+                finishedTasks.poll();
+            } finally {
+                lock.unlock();
+            }
+        }
+        future.setSuccess(ReceiveMessageResponse.builder().messages(messages).build());
+    }
+
+    private void satisfyFuturesFromBuffer() {
+        pruneExpiredTasks();
+        lock.lock();
+        try {
+            futures.forEach(future -> {
+                if (!finishedTasks.isEmpty()) {
+                    fulfillFuture(future, finishedTasks.peek());
+                    futures.remove(future);
+                }
+            });
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void pruneExpiredTasks() {
+        futures.removeIf(ReceiveMessageCompletableFuture::isExpired);
```

Comment:
Should we cancel the future?

Reply:
Nope , in this feature we donot cancel the expired futures we return empty messages via
```java
    private void timeout() {

        if (!responseCompletableFuture.isDone()) {
            responseCompletableFuture.complete(ReceiveMessageResponse.builder().build());
        }
    }
```

This is same as V1 , if SQS is not able to process receive message we need to return empty message
This behaviour is same as SQS service where after longpolling/short polling if SQS doesnot has messages it just sends an empty response.